v8.5.6: dm experimentally supports foreign key (#21523)
pingcap/docs-cn: dm/dm-safe-mode.md

* `INSERT` 语句会被改写成 `REPLACE` 语句。
* `UPDATE` 语句会被分析，得到该语句涉及的行的主键或唯一索引的值，然后改写成 `DELETE` + `REPLACE` 语句 ：先根据主键或唯一索引的定位删除对应的行，然后使用 `REPLACE` 语句插入一条最新值的行记录。

>     从 v8.5.6 起，当你在同步任务的会话中设置 `foreign_key_checks=1` 时，DM 会跳过未修改主键或唯一索引值的 `UPDATE` 语句中的 `DELETE` 步骤。详情参见[外键处理](#外键处理-从-v856-版本开始引入)。

Reviewer: ```suggestion
    从 v8.5.6 起，当你在同步任务的会话中设置 `foreign_key_checks=1` 时，对于未修改主键或唯一索引值的 `UPDATE` 语句，DM 会跳过 `DELETE` 步骤。详情参见[外键处理](#外键处理-从-v856-版本开始引入)。
```

    syncer-config-name: "global"            # syncers 配置的名称
```

## 外键处理 <span class="version-mark">从 v8.5.6 版本开始引入</span>

> **警告：**
>

> > 该功能在 v8.5.6 中引入，目前为实验特性。

Reviewer: ```suggestion
> 该功能为实验特性。不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tiflow/issues) 反馈。
```

当你启用安全模式并在下游任务会话中设置 `foreign_key_checks=1` 时，`UPDATE` 语句默认的 `DELETE` + `REPLACE` 改写方式可能会在子表行上触发非预期的 `ON DELETE CASCADE` 效果。从 v8.5.6 起，DM 引入以下改进来解决此问题。

### 非键值 `UPDATE` 优化

> 对于未修改主键或唯一键值的 `UPDATE` 语句，DM 会跳过 `DELETE` 步骤，仅执行 `REPLACE INTO`。由于主键保持不变，`REPLACE INTO` 会覆盖现有行，而不会触发级联删除。该优化在安全模式下自动生效。

Reviewer: ```suggestion
对于未修改主键或唯一键值的 `UPDATE` 语句，DM 会跳过 `DELETE` 步骤，仅执行 `REPLACE INTO`。由于主键保持不变，`REPLACE INTO` 会覆盖现有行，而不会触发外键级联删除。该优化在安全模式下自动生效。
```

> 例如，考虑以下上游语句，其中 `id` 为主键：

Reviewer: ```suggestion
以下面这个上游语句为例，其中 `id` 为主键：
```

```sql
UPDATE dummydb.dummytbl SET int_value = 888999 WHERE id = 123;
```

在 v8.5.6 之前的版本中，安全模式会将该语句改写为：

```sql
DELETE FROM dummydb.dummytbl WHERE id = 123;       -- 触发 ON DELETE CASCADE
REPLACE INTO dummydb.dummytbl (id, int_value, ...) VALUES (123, 888999, ...);
```

> 从 v8.5.6 开始，安全模式将该语句改写为：

Reviewer: ```suggestion
从 v8.5.6 开始，安全模式会将该语句改写为：
```

```sql
REPLACE INTO dummydb.dummytbl (id, int_value, ...) VALUES (123, 888999, ...);  -- 不触发级联
```

> **警告：**
>

> > 当 `foreign_key_checks=1` 时，DM 不支持修改主键或唯一键值的 `UPDATE` 语句。在这种情况下，任务会暂停，并报错：`safe-mode update with foreign_key_checks=1 and PK/UK changes is not supported`。若需要在包含外键的表上同步此类 `UPDATE` 语句，请将 `safe-mode` 设置为 `false`。

Reviewer: ```suggestion
> 当 `foreign_key_checks=1` 时，DM 不支持同步修改了主键或唯一键值的 `UPDATE` 语句。在这种情况下，同步任务会暂停，并报错：`safe-mode update with foreign_key_checks=1 and PK/UK changes is not supported`。如需在包含外键的表上同步此类 `UPDATE` 语句，请将 `safe-mode` 设置为 `false`。
```

### 会话级 `foreign_key_checks`

> 在安全模式下进行批量执行时，DM 会在执行 `INSERT` 和 `UPDATE` 批处理之前执行 `SET SESSION foreign_key_checks=0`，并在执行完成后恢复原始值。这可以防止 `REPLACE INTO`（其内部实现为 `DELETE` + `INSERT`）在下游触发级联操作。

Reviewer: ```suggestion
在安全模式下执行批处理时，DM 会在执行 `INSERT` 和 `UPDATE` 批处理之前执行 `SET SESSION foreign_key_checks=0`，并在执行完成后恢复 `foreign_key_checks` 原始值。这可以防止 `REPLACE INTO`（其内部实现为 `DELETE` + `INSERT`）在下游触发外键级联操作。
```

该会话级设置在每个批次中会引入少量开销（两次 `SET SESSION` 往返）。在大多数工作负载中，这种开销可以忽略不计。

### 多 worker 外键因果关系

> 当你将 `worker-count` 设置为大于 1 且任务包含带有外键的表时，DM 会在任务启动时从下游的 `CREATE TABLE` schema 中读取外键关系。对于每个 DML 操作，DM 会基于这些关系注入因果键，从而确保父表行及其依赖的子表行的操作被分配到同一个 DML worker 队列中。

Reviewer: ```suggestion
当你将 `worker-count` 设置为大于 1 且同步任务包含带有外键的表时，DM 会在任务启动时从下游的 `CREATE TABLE` schema 中读取外键关系。对于每个 DML 操作，DM 会基于这些关系注入因果键，从而确保父表行及其依赖的子表行的操作被分配到同一个 DML worker 队列中。
```
